File: Old/dynamic-programming/memoization/deleteEarn.cpp
class Solution {
public:
    int n;
    unordered_map<int, int> freq;
    int f(int i, vector<int>& nums, vector<int>& dp){
        if(i == 0)
            return freq[nums[0]]*nums[0];
        if(i < 0)
            return 0;
        
        if(dp[i] != -1) return dp[i];
        
        int notPick = f(i-1, nums, dp);
        int pick = 0;
        if(nums[i] != nums[i-1]+1)
            pick = freq[nums[i]]*nums[i] + f(i - 1, nums, dp);
        else
            pick = freq[nums[i]]*nums[i] + f(i - 2, nums, dp);
        return dp[i] = max (pick,notPick);
    }
    int deleteAndEarn(vector<int>& nums) {
        vector<int> uniqEle;
        for(const auto& num : nums){
            if(freq[num] == 0)
                uniqEle.push_back(num);
            freq[num]++;
        }
        n = uniqEle.size();
        vector<int> dp(n, -1);
        sort(uniqEle.begin(), uniqEle.end());
        return f(n-1, uniqEle, dp);
    }
};
int main(){
  Solution sol = new Solution()
  sol.deleteAndEarn()
  return 0
}

File: Old/dynamic-programming/memoization/deleteEarn.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var deleteAndEarn = function(nums) {
  const sortAscending = (first,second)=> first>second?1:-1
  const frequency = {}
  for(const element of nums){
    if(frequency[element]) {
      frequency[element] += 1
    } else {
      frequency[element] = 1
    }
  }

  const getMaxEarn = (idx,nums,memo={}) => {
    
    
    if (idx in memo) {	
      // console.log(memo[idx],memo	)		
      return memo[idx]
    }
    if (idx === 0){
      return frequency[nums[idx]]*nums[idx]
    }
    if(idx < 0) {
      return 0
    }
    let choice1 = 0
    const currentNumber = nums[idx]
    if(currentNumber !== nums[idx-1] +1 ){
      choice1 = frequency[currentNumber]* currentNumber + getMaxEarn(idx-1,nums,memo)
    } else {
      choice1 = frequency[currentNumber]* currentNumber + getMaxEarn(idx-2,nums,memo)
    }  
    const choice2 = getMaxEarn(idx-1,nums,memo)
    const concluded_choice = Math.max(choice1,choice2)
    // console.log(`choices - ${idx}`,choice1,choice2)
    memo[idx] = concluded_choice
    
    return concluded_choice
  }
  const uniqueNumbers = [...new Set(nums)]
  uniqueNumbers.sort(sortAscending)
  const n = uniqueNumbers.length -1  
  const ans = getMaxEarn(n,uniqueNumbers)

  return ans
  
};
console.log(deleteAndEarn([2,2,3,3,3,4]))